File: app/src/main/java/sugtao4423/icondialog/IconDialog.java
package sugtao4423.icondialog;

import android.annotation.SuppressLint;
import android.app.Activity;
import android.app.AlertDialog;
import android.app.AlertDialog.Builder;
import android.content.Context;
import android.content.DialogInterface.OnClickListener;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ArrayAdapter;
import android.widget.ImageView;
import android.widget.TextView;
import sugtao4423.yandereviewer.R;

public class IconDialog{

	private Builder builder;
	private Context context;

	public IconDialog(Context context){
		builder = new Builder(context);
		this.context = context;
	}

	public Builder setTitle(String title){
		return builder.setTitle(title);
	}

	public Builder setItems(IconItem[] items, OnClickListener listener){
		IconDialogAdapter adapter = new IconDialogAdapter(context, items);
		return builder.setAdapter(adapter, listener);
	}

	public AlertDialog show(){
		return builder.show();
	}
}

class IconDialogAdapter extends ArrayAdapter<IconItem>{

	private LayoutInflater mInflater;

	public IconDialogAdapter(Context context, IconItem[] items){
		super(context, android.R.layout.select_dialog_item, android.R.id.text1, items);
		mInflater = (LayoutInflater)context.getSystemService(Activity.LAYOUT_INFLATER_SERVICE);
	}

	class ViewHolder{
		ImageView image;
		TextView text;
	}

	@SuppressLint("InflateParams")
	@Override
	public View getView(int position, View convertView, ViewGroup parent){
		ViewHolder holder;
		IconItem item = getItem(position);
		if(convertView == null){
			convertView = mInflater.inflate(R.layout.item_dialog_icontext, null);
			holder = new ViewHolder();
			holder.image = (ImageView)convertView.findViewById(R.id.dialog_image);
			holder.text = (TextView)convertView.findViewById(R.id.dialog_text);

			convertView.setTag(holder);
		}else{
			holder = (ViewHolder)convertView.getTag();
		}
		holder.image.setImageResource(item.getResource());
		holder.text.setText(item.getTitle());
		return convertView;
	}
}

File: app/src/main/java/sugtao4423/icondialog/IconDialog.kt
package sugtao4423.icondialog

import android.app.Activity
import android.app.AlertDialog
import android.content.Context
import android.content.DialogInterface
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.ArrayAdapter
import android.widget.ImageView
import android.widget.TextView
import sugtao4423.yandereviewer.R

class IconDialog(private val context: Context) {

    private val builder = AlertDialog.Builder(context)

    fun setTitle(title: String): AlertDialog.Builder {
        return builder.setTitle(title)
    }

    fun setItems(items: Array<IconItem>, listener: DialogInterface.OnClickListener): AlertDialog.Builder {
        val adapter = IconDialogAdapter(context, items)
        return builder.setAdapter(adapter, listener)
    }

    fun show(): AlertDialog {
        return builder.show()
    }

}

class IconDialogAdapter(context: Context, items: Array<IconItem>) :
        ArrayAdapter<IconItem>(context, android.R.layout.select_dialog_item, android.R.id.text1, items) {

    private val mInflater = context.getSystemService(Activity.LAYOUT_INFLATER_SERVICE) as LayoutInflater

    data class ViewHolder(
            val image: ImageView,
            val text: TextView
    )

    override fun getView(position: Int, convertView: View?, parent: ViewGroup): View {
        var view = convertView
        val holder: ViewHolder
        if (view == null) {
            view = mInflater.inflate(R.layout.item_dialog_icontext, parent, false)
            val image = view.findViewById<ImageView>(R.id.dialog_image)
            val text = view.findViewById<TextView>(R.id.dialog_text)

            holder = ViewHolder(image, text)
            view.tag = holder
        } else {
            holder = view.tag as ViewHolder
        }

        val item = getItem(position) ?: return view!!

        holder.image.setImageResource(item.resource)
        holder.text.text = item.title

        return view!!
    }
}